httpcache.go

```diff
@@ -498,6 +516,41 @@ func headerAllCommaSepValues(headers http.Header, name string) []string {
 	return vals
 }
 
+// cachedReadCloser is a wrapper around ReadCloser R that calls OnEOF
+// handler with a full copy of the content read from that ReadCloser
+// when EOF is read.
+type cachedReadCloser struct {
+	// Underlying ReadCloser.
+	R io.ReadCloser
+	// OnEOF is called with a copy of the content of R when EOF is reached.
+	OnEOF func(io.ReadCloser)
+
+	buf bytes.Buffer // buf stores a copy of the content of r.
+}
+
+// Read reads the next len(p) bytes from r or until r is drained. The
+// return value n is the number of bytes read. If r has no data to
+// return, err is io.EOF (unless len(p) is zero) and onEOF is called
+// with a full copy of what has been read so far.
+func (r *cachedReadCloser) Read(p []byte) (n int, err error) {
+	n, err = r.R.Read(p)
+	if n > 0 {
+		if n, err := r.buf.Write(p[:n]); err != nil {
+			return n, err
+		}
+	}
+	if err == io.EOF {
+		buf := r.buf
+		r.OnEOF(ioutil.NopCloser(&buf))
```

Comment:
What you're trying to do here is send a copy of what has been read into `r.buf` so far, in the form of an `io.ReadCloser` (or, as I suggested making it `io.Reader`).

I haven't done any benchmarks, but copying `buf` feels a heavyweight way of achieving that. How about using `bytes.NewReader` instead:

```Go
r.OnEOF(bytes.NewReader(r.buf.Bytes()))
```

From a brief look at implementation details, `bytes.Buffer` has a 64 byte array that would need to be copied. `*bytes.Reader` is a much smaller, more lightweight struct.